hawkular-alerts-engine/src/main/resources/org/hawkular/alerts/engine/rules/ConditionMatch.drl

```diff
@@ -314,44 +317,81 @@ rule ProvideDefaultDampening
         $t : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
         not Dampening( tenantId == $tenantId, triggerId == $tid, triggerMode == $tmode )
     then
-        if (log != null && log.isDebugEnabled()) {
-            log.debug("Adding default " + $tmode + " dampening for trigger! " + $t.getId());
-        }
+        log.debugf("Adding default %s dampening for trigger! %s", $tmode, $t.getId());
+
         Dampening d = Dampening.forStrict( $tenantId, $tid, $tmode, 1 );
         insert( d );
 end
 
-rule CollectConditionEvals
+// If we have multiple ConditionEvals for the same [multi-condition] Trigger, group them together in
+// a TriggerConditionEvalSet that can then be used to update dampening.
+//
+// Note that these two rules will fire before the dampening rules due to complexity and recency. 
+
+rule ProvideTriggerConditionEvalSet
     when
-        $t  : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
-        $ces : TriggerConditionEval( size > 0 )
-               from accumulate(
-                $ce: ConditionEval ( tenantId == $tenantId,  triggerId == $tid ),
-                init( TriggerConditionEval tce = new TriggerConditionEval($tenantId, $tid, $tmode); ),
-                action ( tce.addConditionEval($ce); ),
-                result( tce )
-               )
+        $t   : Trigger( $tenantId : tenantId, $tid : id, $tmode : mode )
+        not TriggerConditionEvalSet( tenantId == $tenantId, triggerId == $tid )
+        $ce1 : ConditionEval ( tenantId == $tenantId,  triggerId == $tid )
+        $ce2 : ConditionEval ( tenantId == $tenantId,  triggerId == $tid, this != $ce1 )
     then
-        $ces.getConditionEvals().stream().forEach(ce -> retract(ce));
-        insert( $ces );
+        retract( $ce1 )
+        retract( $ce2 )
+
+        TriggerConditionEvalSet tces = new TriggerConditionEvalSet( $tenantId, $tid, $ce1, $ce2 );
+        insert( tces );
+
+        log.debugf("ProvideTriggerConditionEvalSet %s", tces);
 end
 
-rule DampenTrigger
+rule AddToTriggerConditionEvalSet

```

Comment:
I don't think there is an issue because addToTriggerConditionEvalSet should match repeatedly due to each firing being against a different ConditionEval.  But I'll take a look now to see if a combined approach may be more clear while maintaining a performance gain.